superdesk/users/async_service.py

```diff
@@ -0,0 +1,537 @@
+def get_admin_privileges():
+    """Get privileges for admin user."""
+    return dict.fromkeys([p["name"] for p in get_privilege_list()], 1)
+
+
+def get_privileges(user, role):
+    """Get privileges for given user and role.
+
+    :param user
+    :param role
+    """
+    if is_admin(user):
+        return get_admin_privileges()
+
+    if role:
+        role_privileges = role.get("privileges", {})
+        return dict(list(role_privileges.items()) + list(user.get("privileges", {}).items()))
+
+    return user.get("privileges", {})
+
+
+def current_user_has_privilege(privilege):
+    """Test if current user has given privilege.
+
+    In case there is no current user we assume it's system (via worker/manage.py)
+    and let it pass.
+
+    :param privilege
+    """
+    if not getattr(g, "user", None):  # no user - worker can do it
+        return True
+    privileges = get_privileges(g.user, getattr(g, "role", None))
+    return privileges.get(privilege, False)
+
+
+def current_user_has_item_privilege(resource: str, item):
+    """Check if current user has privilege for item."""
+    return current_user_has_privilege(get_item_privilege_name(resource, item))
+
+
+def is_sensitive_update(updates):
+    """Test if given update is sensitive and might change user privileges."""
+    return "role" in updates or "privileges" in updates or "user_type" in updates
+
+
+def get_invisible_stages(user_id):
+    user_desks = list(get_resource_service("user_desks").get(req=None, lookup={"user_id": user_id}))
+    user_desk_ids = [d["_id"] for d in user_desks]
+    return get_resource_service("stages").get_stages_by_visibility(False, user_desk_ids)
+
+
+def set_sign_off(user):
+    """
+    Set sign_off property on user if it's not set already.
+    """
+
+    if SIGN_OFF not in user or user[SIGN_OFF] is None:
+        sign_off_mapping = get_app_config("SIGN_OFF_MAPPING", None)
+        if sign_off_mapping and sign_off_mapping in user:
+            user[SIGN_OFF] = user[sign_off_mapping]
+        elif SIGN_OFF in user and user[SIGN_OFF] is None:
+            user[SIGN_OFF] = ""
+        elif "first_name" not in user or "last_name" not in user:
+            user[SIGN_OFF] = user["username"][:3].upper()
+        else:
+            user[SIGN_OFF] = "{first_name[0]}{last_name[0]}".format(**user)
+
+
+def update_sign_off(updates):
+    """
+    Update sign_off property on user if the mapped field is changed.
+    """
+
+    sign_off_mapping = get_app_config("SIGN_OFF_MAPPING", None)
+    if sign_off_mapping and sign_off_mapping in updates:
+        updates[SIGN_OFF] = updates[sign_off_mapping]
+
+
+def get_sign_off(user):
+    """
+    Gets sign_off property on user if it's not set already.
+    """
+
+    if SIGN_OFF not in user or user[SIGN_OFF] is None:
+        set_sign_off(user)
+
+    return user[SIGN_OFF]
+
+
+class UsersAsyncService(AsyncResourceService[UsersResourceModel]):
+    _updating_stage_visibility = True
+
+    async def __is_invalid_operation(
```

Comment:
After further inspection, looks like this one is not needed here either, as it is in the context of an API request.

It's functions like these should not go into these new async resource services, but instead the REST API (when we implement the User one).

We want to keep the separation of concerns with the new code, by separating out the data and request layers specifically. This should help keep these two concerns smaller, easier to read etc

Reply:
@MarkLark86 for that function, as it is currently used in the `on_update` and `on_delete` functions in the service, do I remove it still in those function as it is going to be handled in the REST API before the functions are called ?

Reply:
Correct, we will validate the request in the REST API code